### DISCUSSION TITLE
Search reverse wildcard postfix doesn't work the same as wildcard prefix

### DISCUSSION THREAD

**Discussion by JoshClose** [ravendb/ravendb#18961] [Q&A] [answered]
The documentation says to use a reverse wildcard postfix instead of prefix and postfix. They don't seem to work the same for me.

When doing a reverse search, it will only match the beginning or end of the field.
```
where search(Field, "123*") 
or search(FieldReverse, "321*")
```
```
123b - matches normal
a123 - matches reverse
a123b - never matches
```

When doing a prefix and postfix search, it will match text in the middle of the field.
```
where search(Field, "*foo*")
```
```
123b - matches normal
a123 - matches reverse
a123b - matches in middle
```

Is this expected behavior or am I doing something wrong here? Based on how I would think prefix and postfix work, these seems like it's working as expected, though this isn't a good alternative for matching in the middle of a word.

**Comment by ayende** [comment]
Can you share the index & data.
From our perspective, assuming that `FieldReverse = doc.Field.Reverse()` - there should be identical behavior.

**Reply by ayende** [reply]
Node that depending on your actual need, ngram may be better suited for you

**Reply by JoshClose** [reply]
The regex is because they are formatted like `"AD-FA-F9-AF-AA-80"`.
```cs
public class AccessPoints_Search : AbstractIndexCreationTask<AccessPoint, AccessPoints_Search.Result>
{
	public class Result
	{
		public string Name { get; set; } = string.Empty;

		public string NameReverse { get; set; } = string.Empty;

		public string MacAddress { get; set; } = string.Empty;

		public string MacAddressReverse { get; set; } = string.Empty;

		public string OrganizationId { get; set; } = string.Empty;

		public string LocationId { get; set; } = string.Empty;
	}

	public AccessPoints_Search()
	{
		Map = accessPoints =>
			from ap in accessPoints
			let nameReverse = ap.Name.Reverse().ToString()
			let mac = Regex.Replace(ap.MacAddress, "[^0-9a-fA-F]", string.Empty)
			let macReverse = Regex.Replace(new string(((string)ap.MacAddress).Reverse().ToArray()), "[^0-9a-fA-F]", string.Empty)
			select new Result
			{
				Name = ap.Name,
				NameReverse = nameReverse,
				MacAddress = mac,
				MacAddressReverse = macReverse,
				OrganizationId = ap.Organization.Id!,
				LocationId = ap.Location.Id!,
			};

		Index(x => x.Name, FieldIndexing.Search);
		Index(x => x.NameReverse, FieldIndexing.Search);
		Index(x => x.MacAddress, FieldIndexing.Search);
		Index(x => x.MacAddressReverse, FieldIndexing.Search);
	}
}
```

**Reply by ayende** [reply]
Start with: `let macReverse = mac.Reverse()` and that should work

I think that your output is different there.
Look at the index terms, and you can see the differences

**Reply by JoshClose** [reply]
Here is the index output from a query.
<img width="887" alt="image" src="https://github.com/user-attachments/assets/f9680a03-7675-4226-b535-eb88327f2b53">
When using the commented out code, only `test2` shows up.

**Reply by ayende** [reply, marked as answer]
Look at the `MacAddressReverse` 

Your search is basically `startsWith(MacAddressReverse, "AFDA")` but you don't _have_ that.
When you are doing reverse, you need to reverse this as well, so search for: `fdfd` and you'll a result.

Assuming I got your point properly, user inputs "94da", you search:

`search(MacAddress, "94da") or search(MacAddressReverse, "ad49")`

**Reply by JoshClose** [reply]
If I have a field of `"123456"`, I want to search for `"34"` and have it return results. All the similar characters in a MAC address make that confusing.

**Reply by ayende** [reply]
You want a `contains` - just use that.
What you are trying to do with the prefix & postfix and reverse is a really complex and not suitable if you _really_ need contains.

You can also use:

```
let chars = ap.MacAddress.Where(char.IsLetterOrDigit).ToArray();
let parts = Enumerable.Range(0, chars.Length/2).Select(x => new string(chars, 0, x*2)).ToArray();
```

In which case the terms you'll index for `00:9A:CD:3C:4D:5E` will be:

```
00
009A
009ACD
009ACD3C
009ACD3C4D
```

**Reply by JoshClose** [reply]
Oh nice! I'll give that a try.

I suppose since the text doesn't have multiple words in it, full-text search probably isn't the right solution.

**Reply by ayende** [reply]
correct
